File: src/features/scheduler/domain/assignment.ts
import { buildDayOverviews } from "./calendar";
import {
  DayOverview,
  DaySchedule,
  MemberAvailability,
  SchedulerInput,
  SchedulerMember,
  SchedulerShiftRequest,
  ShiftType,
} from "./types";

function makeSeed(input: string) {
  let seed = 0;
  for (const char of input) {
    seed = (seed * 31 + char.charCodeAt(0)) >>> 0;
  }
  return seed || 1;
}

function shuffle<T>(items: T[], seedInput: string) {
  const result = items.slice();
  let seed = makeSeed(seedInput);

  for (let index = result.length - 1; index > 0; index -= 1) {
    seed = (seed * 1664525 + 1013904223) >>> 0;
    const swapIndex = seed % (index + 1);
    [result[index], result[swapIndex]] = [result[swapIndex], result[index]];
  }

  return result;
}

function previousDate(date: string) {
  const value = new Date(`${date}T00:00:00`);
  value.setDate(value.getDate() - 1);
  return value.toISOString().slice(0, 10);
}

interface SchedulerState {
  nameByMember: Map<string, string>;
  memberById: Map<string, SchedulerMember>;
  assignmentsByMember: Map<string, Set<string>>;
  assignmentsByShift: Map<string, [string | null, string | null]>;
  assignmentCountByMember: Map<string, number>;
}

interface RankedCandidate {
  memberId: string;
  assignmentCount: number;
  tieBreaker: number;
}

interface CandidatePair {
  pair: [string, string];
  preferenceScore: number;
  fairnessScore: number;
  rankScore: number;
}

function emptyAssignedSlots(): [string | null, string | null] {
  return [null, null];
}

function filledSlotCount(assigned: [string | null, string | null]) {
  return assigned.filter(Boolean).length;
}

function assignedMemberIdsForShift(state: SchedulerState, shiftId: string) {
  return (state.assignmentsByShift.get(shiftId) ?? emptyAssignedSlots()).filter(
    (memberId): memberId is string => Boolean(memberId),
  );
}

function buildSchedulerState(input: {
  availabilities: MemberAvailability[];
  schedulerInput?: SchedulerInput;
}): SchedulerState {
  const members = input.schedulerInput
    ? input.schedulerInput.members
    : input.availabilities.map((availability) => ({
        memberId: availability.memberId,
        name: availability.name,
        email: availability.email,
        active: true,
        maxShifts: null,
        preferredCoworkerIds: [],
        availabilityDates: Array.from(new Set(availability.entries.map((entry) => entry.date))).sort((a, b) =>
          a.localeCompare(b),
        ),
      }));
  const nameByMember = new Map(
    members.map((member) => [member.memberId, member.name]),
  );
  const memberById = new Map(members.map((member) => [member.memberId, member]));
  const assignmentCountByMember = new Map(members.map((member) => [member.memberId, 0]));

  return {
    nameByMember,
    memberById,
    assignmentsByMember: new Map(),
    assignmentsByShift: new Map(),
    assignmentCountByMember,
  };
}

function canTakeShift(
  memberId: string,
  date: string,
  shiftType: ShiftType,
  state: SchedulerState,
) {
  const member = state.memberById.get(memberId);
  if (!member || !member.active) {
    return false;
  }

  const assignmentCount = state.assignmentCountByMember.get(memberId) ?? 0;
  if (member.maxShifts !== null && assignmentCount >= member.maxShifts) {
    return false;
  }

  const existing = state.assignmentsByMember.get(memberId) ?? new Set<string>();

  if (existing.has(`${date}:night`) || existing.has(`${date}:day`)) {
    return false;
  }

  if (shiftType === "day" && existing.has(`${previousDate(date)}:night`)) {
    return false;
  }

  return true;
}

function rankShiftCandidates(monthId: string, shift: SchedulerShiftRequest, state: SchedulerState) {
  const shuffled = shuffle(shift.candidateMemberIds, `${monthId}:${shift.shiftId}:${shift.candidateMemberIds.join(",")}`);

  return shuffled
    .filter((memberId) => state.memberById.has(memberId))
    .map((memberId, index) => ({
      memberId,
      assignmentCount: state.assignmentCountByMember.get(memberId) ?? 0,
      tieBreaker: index,
    }))
    .sort((left, right) => {
      if (left.assignmentCount !== right.assignmentCount) {
        return left.assignmentCount - right.assignmentCount;
      }

      return left.tieBreaker - right.tieBreaker;
    });
}

function findBestSingleCandidateForShift(
  monthId: string,
  shift: SchedulerShiftRequest,
  state: SchedulerState,
  excludedMemberIds: Set<string>,
) {
  const rankedCandidates = rankShiftCandidates(monthId, shift, state);

  return (
    rankedCandidates.find(
      (candidate) =>
        !excludedMemberIds.has(candidate.memberId) &&
        canTakeShift(candidate.memberId, shift.date, shift.shiftType, state),
    )?.memberId ?? null
  );
}

function preferenceScoreForPair(memberA: SchedulerMember, memberB: SchedulerMember) {
  const aPrefersB = memberA.preferredCoworkerIds.includes(memberB.memberId);
  const bPrefersA = memberB.preferredCoworkerIds.includes(memberA.memberId);

  if (aPrefersB && bPrefersA) return 2;
  if (aPrefersB || bPrefersA) return 1;
  return 0;
}

function materializeAssignedNames(
  state: SchedulerState,
  assignedMemberIds: [string | null, string | null],
): [string | null, string | null] {
  return [
    assignedMemberIds[0] ? state.nameByMember.get(assignedMemberIds[0]) ?? null : null,
    assignedMemberIds[1] ? state.nameByMember.get(assignedMemberIds[1]) ?? null : null,
  ];
}

function chooseBestPair(candidates: RankedCandidate[], state: SchedulerState) {
  let bestPair: CandidatePair | null = null;

  for (let leftIndex = 0; leftIndex < candidates.length; leftIndex += 1) {
    for (let rightIndex = leftIndex + 1; rightIndex < candidates.length; rightIndex += 1) {
      const left = candidates[leftIndex];
      const right = candidates[rightIndex];
      const leftMember = state.memberById.get(left.memberId);
      const rightMember = state.memberById.get(right.memberId);
      if (!leftMember || !rightMember) continue;

      const pairScore: CandidatePair = {
        pair: [left.memberId, right.memberId],
        preferenceScore: preferenceScoreForPair(leftMember, rightMember),
        fairnessScore: left.assignmentCount + right.assignmentCount,
        rankScore: left.tieBreaker + right.tieBreaker,
      };

      if (
        !bestPair ||
        pairScore.fairnessScore < bestPair.fairnessScore ||
        (pairScore.fairnessScore === bestPair.fairnessScore &&
          pairScore.preferenceScore > bestPair.preferenceScore) ||
        (pairScore.fairnessScore === bestPair.fairnessScore &&
          pairScore.preferenceScore === bestPair.preferenceScore &&
          pairScore.rankScore < bestPair.rankScore)
      ) {
        bestPair = pairScore;
      }
    }
  }

  return bestPair?.pair ?? null;
}

function assignMemberToShift(state: SchedulerState, shift: SchedulerShiftRequest, memberId: string, slotIndex: 0 | 1) {
  const currentAssignments = state.assignmentsByShift.get(shift.shiftId) ?? emptyAssignedSlots();
  currentAssignments[slotIndex] = memberId;
  state.assignmentsByShift.set(shift.shiftId, currentAssignments);

  const memberAssignments = state.assignmentsByMember.get(memberId) ?? new Set<string>();
  memberAssignments.add(`${shift.date}:${shift.shiftType}`);
  state.assignmentsByMember.set(memberId, memberAssignments);
  state.assignmentCountByMember.set(memberId, (state.assignmentCountByMember.get(memberId) ?? 0) + 1);
}

function removeMemberFromShift(state: SchedulerState, shiftId: string, memberId: string) {
  const currentAssignments = state.assignmentsByShift.get(shiftId) ?? emptyAssignedSlots();
  const nextAssignments = currentAssignments.filter((assignedMemberId) => assignedMemberId && assignedMemberId !== memberId);
  const compacted: [string | null, string | null] = [nextAssignments[0] ?? null, nextAssignments[1] ?? null];
  state.assignmentsByShift.set(shiftId, compacted);

  const memberAssignments = state.assignmentsByMember.get(memberId) ?? new Set<string>();
  memberAssignments.delete(shiftId);
  state.assignmentsByMember.set(memberId, memberAssignments);
  state.assignmentCountByMember.set(memberId, Math.max(0, (state.assignmentCountByMember.get(memberId) ?? 1) - 1));
}

function fillShiftSlots(monthId: string, shift: SchedulerShiftRequest, state: SchedulerState) {
  const rankedCandidates = rankShiftCandidates(monthId, shift, state);
  const validCandidates = rankedCandidates.filter((candidate) =>
    canTakeShift(candidate.memberId, shift.date, shift.shiftType, state),
  );

  if (validCandidates.length === 0) {
    return emptyAssignedSlots();
  }

  if (validCandidates.length === 1) {
    assignMemberToShift(state, shift, validCandidates[0].memberId, 0);
    return state.assignmentsByShift.get(shift.shiftId) ?? emptyAssignedSlots();
  }

  const preferredPair = chooseBestPair(validCandidates, state);
  if (preferredPair) {
    assignMemberToShift(state, shift, preferredPair[0], 0);
    assignMemberToShift(state, shift, preferredPair[1], 1);
    return state.assignmentsByShift.get(shift.shiftId) ?? emptyAssignedSlots();
  }

  let slotIndex: 0 | 1 = 0;
  for (const candidate of validCandidates) {
    assignMemberToShift(state, shift, candidate.memberId, slotIndex);
    if (slotIndex === 1) break;
    slotIndex = 1;
  }

  return state.assignmentsByShift.get(shift.shiftId) ?? emptyAssignedSlots();
}

function buildShiftRequestMap(schedulerInput: SchedulerInput) {
  return new Map(schedulerInput.shiftRequests.map((shift) => [shift.shiftId, shift]));
}

function prioritizeShiftsForRebalance(schedulerInput: SchedulerInput, state: SchedulerState) {
  return schedulerInput.shiftRequests
    .map((shift) => ({
      shift,
      filledSlots: filledSlotCount(state.assignmentsByShift.get(shift.shiftId) ?? emptyAssignedSlots()),
      candidateCount: shift.candidateMemberIds.length,
    }))
    .sort((left, right) => {
      if (left.filledSlots !== right.filledSlots) {
        return left.filledSlots - right.filledSlots;
      }

      if (left.candidateCount !== right.candidateCount) {
        return left.candidateCount - right.candidateCount;
      }

      return left.shift.shiftId.localeCompare(right.shift.shiftId);
    });
}

function tryRebalanceCandidate(
  monthId: string,
  targetShift: SchedulerShiftRequest,
  candidateId: string,
  state: SchedulerState,
  shiftRequests: Map<string, SchedulerShiftRequest>,
) {
  const assignedShiftIds = Array.from(state.assignmentsByMember.get(candidateId) ?? []);
  const targetFilledBefore = filledSlotCount(state.assignmentsByShift.get(targetShift.shiftId) ?? emptyAssignedSlots());

  for (const donorShiftId of assignedShiftIds) {
    if (donorShiftId === targetShift.shiftId) continue;

    const donorShift = shiftRequests.get(donorShiftId);
    if (!donorShift) continue;

    const donorAssignments = state.assignmentsByShift.get(donorShiftId) ?? emptyAssignedSlots();
    const donorFilledBefore = filledSlotCount(donorAssignments);
    if (donorFilledBefore < 2) continue;
    const donorSlotIndex = donorAssignments[0] === candidateId ? 0 : 1;

    removeMemberFromShift(state, donorShiftId, candidateId);

    if (canTakeShift(candidateId, targetShift.date, targetShift.shiftType, state)) {
      const targetAssignments = state.assignmentsByShift.get(targetShift.shiftId) ?? emptyAssignedSlots();
      const slotIndex: 0 | 1 = targetAssignments[0] === null ? 0 : 1;
      assignMemberToShift(state, targetShift, candidateId, slotIndex);

       const donorAssignedMemberIds = new Set(assignedMemberIdsForShift(state, donorShiftId));
       donorAssignedMemberIds.add(candidateId);
       const donorReplacement = findBestSingleCandidateForShift(monthId, donorShift, state, donorAssignedMemberIds);
       if (donorReplacement) {
         assignMemberToShift(state, donorShift, donorReplacement, donorSlotIndex);
         return true;
       }

       if (targetFilledBefore === 0) {
         return true;
       }

       removeMemberFromShift(state, targetShift.shiftId, candidateId);
       assignMemberToShift(state, donorShift, candidateId, donorSlotIndex);
       continue;
    }

    assignMemberToShift(
      state,
      donorShift,
      candidateId,
      donorSlotIndex,
    );
  }

  return false;
}

function rebalanceUnderfilledShifts(
  monthId: string,
  schedulerInput: SchedulerInput,
  state: SchedulerState,
  shiftRequests: Map<string, SchedulerShiftRequest>,
) {
  const prioritizedShifts = prioritizeShiftsForRebalance(schedulerInput, state);

  for (const target of prioritizedShifts) {
    const targetShift = target.shift;
    let targetAssignments = state.assignmentsByShift.get(targetShift.shiftId) ?? emptyAssignedSlots();

    while (filledSlotCount(targetAssignments) < 2) {
      const rankedCandidates = rankShiftCandidates(monthId, targetShift, state);
      const openSlotFilled = rankedCandidates.some((candidate) =>
        tryRebalanceCandidate(monthId, targetShift, candidate.memberId, state, shiftRequests),
      );

      if (!openSlotFilled) break;
      targetAssignments = state.assignmentsByShift.get(targetShift.shiftId) ?? emptyAssignedSlots();
    }
  }
}

function buildFallbackSchedulerInput(
  monthId: string,
  intakeLimitPerShift: number,
  availabilities: MemberAvailability[],
): SchedulerInput {
  const days = buildDayOverviews(monthId, intakeLimitPerShift, availabilities);

  return {
    monthId,
    members: availabilities.map((availability) => ({
      memberId: availability.memberId,
      name: availability.name,
      email: availability.email,
      active: true,
      maxShifts: null,
      preferredCoworkerIds: [],
      availabilityDates: Array.from(new Set(availability.entries.map((entry) => entry.date))).sort((a, b) =>
        a.localeCompare(b),
      ),
    })),
    shiftRequests: days.flatMap((day) =>
      day.shifts.map((shift) => ({
        shiftId: `${day.date}:${shift.shiftType}`,
        date: day.date,
        shiftType: shift.shiftType,
        capacity: shift.capacity,
        candidateMemberIds: shift.candidateIds.slice(),
      })),
    ),
  };
}

function assignInitialSchedule(
  days: DayOverview[],
  shiftRequests: Map<string, SchedulerShiftRequest>,
  state: SchedulerState,
  monthId: string,
) {
  for (const day of days) {
    for (const shift of day.shifts) {
      const shiftRequest = shiftRequests.get(`${day.date}:${shift.shiftType}`);
      if (!shiftRequest) continue;
      fillShiftSlots(monthId, shiftRequest, state);
    }
  }
}

function buildScheduleRows(days: DayOverview[], state: SchedulerState): DaySchedule[] {
  return days.map((day) => {
    const nightAssigned = materializeAssignedNames(
      state,
      state.assignmentsByShift.get(`${day.date}:night`) ?? emptyAssignedSlots(),
    );
    const dayAssigned = materializeAssignedNames(
      state,
      state.assignmentsByShift.get(`${day.date}:day`) ?? emptyAssignedSlots(),
    );

    return {
      date: day.date,
      weekdayLabel: day.weekdayLabel,
      weekend: day.weekend,
      night: nightAssigned,
      day: dayAssigned,
    };
  });
}

export function generateSchedule(
  input:
    | {
        monthId: string;
        intakeLimitPerShift: number;
        availabilities: MemberAvailability[];
      }
    | {
        monthId: string;
        intakeLimitPerShift: number;
        availabilities: MemberAvailability[];
        schedulerInput: SchedulerInput;
      },
): DaySchedule[] {
  const { monthId, intakeLimitPerShift, availabilities } = input;
  const schedulerInput = "schedulerInput" in input
    ? input.schedulerInput
    : buildFallbackSchedulerInput(monthId, intakeLimitPerShift, availabilities);
  const state = buildSchedulerState({
    availabilities,
    schedulerInput,
  });
  const shiftRequests = buildShiftRequestMap(schedulerInput);
  const days = buildDayOverviews(monthId, intakeLimitPerShift, availabilities);
  assignInitialSchedule(days, shiftRequests, state, monthId);
  rebalanceUnderfilledShifts(monthId, schedulerInput, state, shiftRequests);
  return buildScheduleRows(days, state);
}

File: backend/app/domain/assignment.py
from __future__ import annotations
from datetime import date, timedelta

from app.domain.calendar import build_day_overviews
from app.domain.types import (
    DayOverview,
    DaySchedule,
    MemberAvailability,
    SchedulerInput,
    SchedulerMember,
    SchedulerShiftRequest,
    ShiftType,
)


# ---------------------------------------------------------------------------
# Deterministic shuffle (mirrors the TS LCG implementation exactly)
# ---------------------------------------------------------------------------

def _make_seed(s: str) -> int:
    seed = 0
    for ch in s:
        seed = ((seed * 31) + ord(ch)) & 0xFFFFFFFF
    return seed or 1


def _shuffle(items: list, seed_input: str) -> list:
    result = list(items)
    seed = _make_seed(seed_input)
    for i in range(len(result) - 1, 0, -1):
        seed = ((seed * 1664525) + 1013904223) & 0xFFFFFFFF
        j = seed % (i + 1)
        result[i], result[j] = result[j], result[i]
    return result


def _previous_date(date_str: str) -> str:
    d = date.fromisoformat(date_str)
    return (d - timedelta(days=1)).isoformat()


# ---------------------------------------------------------------------------
# Scheduler state
# ---------------------------------------------------------------------------

class _State:
    def __init__(self, members: list[SchedulerMember]) -> None:
        self.name_by_member: dict[str, str] = {m.member_id: m.name for m in members}
        self.member_by_id: dict[str, SchedulerMember] = {m.member_id: m for m in members}
        self.assignments_by_member: dict[str, set[str]] = {}
        self.assignments_by_shift: dict[str, list[str | None]] = {}
        self.count_by_member: dict[str, int] = {m.member_id: 0 for m in members}


def _empty_slots() -> list[str | None]:
    return [None, None]


def _filled_count(slots: list[str | None]) -> int:
    return sum(1 for s in slots if s)


def _assigned_ids_for_shift(state: _State, shift_id: str) -> list[str]:
    return [s for s in state.assignments_by_shift.get(shift_id, _empty_slots()) if s]


def _can_take_shift(member_id: str, date_str: str, shift_type: ShiftType, state: _State) -> bool:
    member = state.member_by_id.get(member_id)
    if not member or not member.active:
        return False
    if member.max_shifts is not None and state.count_by_member.get(member_id, 0) >= member.max_shifts:
        return False
    assigned = state.assignments_by_member.get(member_id, set())
    if f"{date_str}:night" in assigned or f"{date_str}:day" in assigned:
        return False
    if shift_type == "day" and f"{_previous_date(date_str)}:night" in assigned:
        return False
    return True


def _rank_candidates(month_id: str, shift: SchedulerShiftRequest, state: _State) -> list[dict]:
    shuffled = _shuffle(shift.candidate_member_ids, f"{month_id}:{shift.shift_id}:{','.join(shift.candidate_member_ids)}")
    ranked = [
        {"member_id": mid, "count": state.count_by_member.get(mid, 0), "tie": i}
        for i, mid in enumerate(shuffled)
        if mid in state.member_by_id
    ]
    return sorted(ranked, key=lambda c: (c["count"], c["tie"]))


def _assign(state: _State, shift: SchedulerShiftRequest, member_id: str, slot_index: int) -> None:
    slots = state.assignments_by_shift.setdefault(shift.shift_id, _empty_slots())
    slots[slot_index] = member_id
    state.assignments_by_member.setdefault(member_id, set()).add(f"{shift.date}:{shift.shift_type}")
    state.count_by_member[member_id] = state.count_by_member.get(member_id, 0) + 1


def _remove(state: _State, shift_id: str, member_id: str) -> None:
    slots = state.assignments_by_shift.get(shift_id, _empty_slots())
    compacted: list[str | None] = [s for s in slots if s and s != member_id]
    state.assignments_by_shift[shift_id] = [compacted[0] if len(compacted) > 0 else None,
                                             compacted[1] if len(compacted) > 1 else None]
    assigned = state.assignments_by_member.get(member_id, set())
    assigned.discard(shift_id)
    state.count_by_member[member_id] = max(0, state.count_by_member.get(member_id, 1) - 1)


def _preference_score(a: SchedulerMember, b: SchedulerMember) -> int:
    a_prefers_b = b.member_id in a.preferred_coworker_ids
    b_prefers_a = a.member_id in b.preferred_coworker_ids
    if a_prefers_b and b_prefers_a:
        return 2
    if a_prefers_b or b_prefers_a:
        return 1
    return 0


def _choose_best_pair(candidates: list[dict], state: _State) -> tuple[str, str] | None:
    best = None
    for i in range(len(candidates)):
        for j in range(i + 1, len(candidates)):
            left, right = candidates[i], candidates[j]
            lm = state.member_by_id.get(left["member_id"])
            rm = state.member_by_id.get(right["member_id"])
            if not lm or not rm:
                continue
            fairness = left["count"] + right["count"]
            pref = _preference_score(lm, rm)
            rank = left["tie"] + right["tie"]
            if best is None:
                best = (left["member_id"], right["member_id"], fairness, pref, rank)
            else:
                _, _, bf, bp, br = best
                if (fairness < bf
                        or (fairness == bf and pref > bp)
                        or (fairness == bf and pref == bp and rank < br)):
                    best = (left["member_id"], right["member_id"], fairness, pref, rank)
    return (best[0], best[1]) if best else None


def _find_best_single(
    month_id: str, shift: SchedulerShiftRequest, state: _State, excluded: set[str]
) -> str | None:
    for c in _rank_candidates(month_id, shift, state):
        if c["member_id"] not in excluded and _can_take_shift(c["member_id"], shift.date, shift.shift_type, state):
            return c["member_id"]
    return None


def _fill_shift_slots(month_id: str, shift: SchedulerShiftRequest, state: _State) -> None:
    ranked = [c for c in _rank_candidates(month_id, shift, state)
              if _can_take_shift(c["member_id"], shift.date, shift.shift_type, state)]

    if not ranked:
        return
    if len(ranked) == 1:
        _assign(state, shift, ranked[0]["member_id"], 0)
        return

    pair = _choose_best_pair(ranked, state)
    if pair:
        _assign(state, shift, pair[0], 0)
        _assign(state, shift, pair[1], 1)
        return

    for idx, c in enumerate(ranked[:2]):
        _assign(state, shift, c["member_id"], idx)


def _try_rebalance(
    month_id: str,
    target: SchedulerShiftRequest,
    candidate_id: str,
    state: _State,
    shift_map: dict[str, SchedulerShiftRequest],
) -> bool:
    assigned_shifts = list(state.assignments_by_member.get(candidate_id, set()))
    target_filled_before = _filled_count(state.assignments_by_shift.get(target.shift_id, _empty_slots()))

    for donor_id in assigned_shifts:
        if donor_id == target.shift_id:
            continue
        donor = shift_map.get(donor_id)
        if not donor:
            continue
        donor_slots = state.assignments_by_shift.get(donor_id, _empty_slots())
        if _filled_count(donor_slots) < 2:
            continue
        donor_slot_index = 0 if donor_slots[0] == candidate_id else 1

        _remove(state, donor_id, candidate_id)

        if _can_take_shift(candidate_id, target.date, target.shift_type, state):
            target_slots = state.assignments_by_shift.get(target.shift_id, _empty_slots())
            slot_index = 0 if target_slots[0] is None else 1
            _assign(state, target, candidate_id, slot_index)

            already_in_donor = set(_assigned_ids_for_shift(state, donor_id))
            already_in_donor.add(candidate_id)
            replacement = _find_best_single(month_id, donor, state, already_in_donor)
            if replacement:
                _assign(state, donor, replacement, donor_slot_index)
                return True
            if target_filled_before == 0:
                return True
            _remove(state, target.shift_id, candidate_id)

        _assign(state, donor, candidate_id, donor_slot_index)

    return False


def _rebalance(month_id: str, scheduler_input: SchedulerInput, state: _State, shift_map: dict) -> None:
    prioritized = sorted(
        scheduler_input.shift_requests,
        key=lambda s: (
            _filled_count(state.assignments_by_shift.get(s.shift_id, _empty_slots())),
            len(s.candidate_member_ids),
            s.shift_id,
        ),
    )

    for target in prioritized:
        while _filled_count(state.assignments_by_shift.get(target.shift_id, _empty_slots())) < 2:
            ranked = _rank_candidates(month_id, target, state)
            improved = any(
                _try_rebalance(month_id, target, c["member_id"], state, shift_map) for c in ranked
            )
            if not improved:
                break


def _build_fallback_input(month_id: str, intake_limit: int, availabilities: list[MemberAvailability]) -> SchedulerInput:
    from app.domain.types import SchedulerInput
    from app.domain.scheduler_input import build_scheduler_shift_requests
    days = build_day_overviews(month_id, intake_limit, availabilities)
    members = [
        SchedulerMember(
            member_id=a.member_id,
            name=a.name,
            email=a.email,
            active=True,
            max_shifts=None,
            preferred_coworker_ids=[],
            availability_dates=sorted(set(e.date for e in a.entries)),
        )
        for a in availabilities
    ]
    return SchedulerInput(
        month_id=month_id,
        members=members,
        shift_requests=build_scheduler_shift_requests(month_id, intake_limit, availabilities),
    )


def generate_schedule(
    month_id: str,
    intake_limit_per_shift: int,
    availabilities: list[MemberAvailability],
    scheduler_input: SchedulerInput | None = None,
) -> list[DaySchedule]:
    if scheduler_input is None:
        scheduler_input = _build_fallback_input(month_id, intake_limit_per_shift, availabilities)

    state = _State(scheduler_input.members)
    shift_map = {s.shift_id: s for s in scheduler_input.shift_requests}
    days = build_day_overviews(month_id, intake_limit_per_shift, availabilities)

    for day in days:
        for shift in day.shifts:
            sr = shift_map.get(f"{day.date}:{shift.shift_type}")
            if sr:
                _fill_shift_slots(month_id, sr, state)

    _rebalance(month_id, scheduler_input, state, shift_map)

    result: list[DaySchedule] = []
    for day in days:
        night_slots = state.assignments_by_shift.get(f"{day.date}:night", _empty_slots())
        day_slots = state.assignments_by_shift.get(f"{day.date}:day", _empty_slots())
        result.append(DaySchedule(
            date=day.date,
            weekday_label=day.weekday_label,
            weekend=day.weekend,
            night=(
                state.name_by_member.get(night_slots[0]) if night_slots[0] else None,
                state.name_by_member.get(night_slots[1]) if night_slots[1] else None,
            ),
            day=(
                state.name_by_member.get(day_slots[0]) if day_slots[0] else None,
                state.name_by_member.get(day_slots[1]) if day_slots[1] else None,
            ),
            night_ids=(night_slots[0], night_slots[1]),
            day_ids=(day_slots[0], day_slots[1]),
        ))

    return result
